typescript/sdk/src/ism/HyperlaneIsmFactory.ts

```diff
@@ -0,0 +1,250 @@
+          .map((log) => {
+            try {
+              return routingIsmFactory.interface.parseLog(log);
+            } catch (e) {
+              return undefined;
+            }
+          })
+          .filter(
+            (log): log is ethers.utils.LogDescription =>
+              !!log && log.name === 'ModuleDeployed',
+          );
+        const moduleAddress = dispatchLogs[0].args['module'];
+        const routingIsm = DomainRoutingIsm__factory.connect(
+          moduleAddress,
+          this.multiProvider.getSigner(chain),
+        );
+        await routingIsm.transferOwnership(config.owner);
+        const address = dispatchLogs[0].args['module'];
+        return DomainRoutingIsm__factory.connect(address, signer);
+      }
+      case ModuleType.AGGREGATION: {
+        const aggregationIsmFactory =
+          this.getContracts(chain).aggregationIsmFactory;
+        const addresses: types.Address[] = [];
+        for (const module of config.modules) {
+          addresses.push((await this.deploy(chain, module)).address);
+        }
+        const address = await this.deployMOfNFactory(
+          chain,
+          aggregationIsmFactory,
+          addresses,
+          config.threshold,
+        );
+        return StaticAggregationIsm__factory.connect(address, signer);
+      }
+      default: {
+        throw new Error('Unknown ModuleType');
+      }
+    }
+  }
+
+  private async deployMOfNFactory(
+    chain: ChainName,
+    factory: StaticMOfNAddressSetFactory,
+    values: types.Address[],
+    threshold: number,
+  ): Promise<types.Address> {
+    const address = await factory.getAddress(values.sort(), threshold);
+    const provider = this.multiProvider.getProvider(chain);
+    const code = await provider.getCode(address);
+    if (code === '0x') {
+      await factory.deploy(values.sort(), threshold);
+    }
+    return address;
+  }
+}
+
+export async function moduleMatches(
+  chain: ChainName,
+  moduleAddress: types.Address,
+  config: IsmConfig,
+  multiProvider: MultiProvider,
+  contracts: HyperlaneContracts<IsmFactoryFactories>,
+): Promise<boolean> {
+  const provider = multiProvider.getProvider(chain);
+  const module = IInterchainSecurityModule__factory.connect(
+    moduleAddress,
+    provider,
+  );
+  const actualType = await module.moduleType();
+  if (actualType !== config.type) return false;
+  let matches = true;
+  switch (config.type) {
+    case ModuleType.MULTISIG: {
+      // A MultisigIsm matches if validators and threshold match the config
+      const expectedAdddress = await contracts.multisigIsmFactory.getAddress(
+        config.validators.sort(),
+        config.threshold,
+      );
+      matches = eqAddress(expectedAdddress, module.address);
+      break;
+    }
+    case ModuleType.ROUTING: {
+      // A RoutingIsm matches if:
+      //   1. The set of domains in the config equals those on-chain
+      //   2. The modules for each domain match the config
+      // TODO: Check (1)
```

Comment:
resolve?

Reply:
So we're not actually checking this yet. I'm not sure exactly what to do. I could make the DomainRoutingIsm use an EnumerableMap, and expose a `domains()` function, but I'm not sure that should be part of the general RoutingIsm interface.

So eventually we'll need to move to a point where we use bytecode or st to check which implementation of a particular interface we're looking at. And ofc not all implementations will be known to us.